### PR TITLE
feat(base): nullifier tree impl

### PR DESCRIPTION
**Pull request by cheethas**
# Description
previous branch got merked

Please provide a paragraph or two giving a summary of the change, including relevant motivation and context.

# Checklist:

- [ ] I have reviewed my diff in github, line by line.
- [ ] Every change is related to the PR description.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this pull request to the issue(s) that it resolves.
- [ ] There are no unexpected formatting changes, superfluous debug logs, or commented-out code.
- [ ] There are no circuit changes, OR specifications in `/specs` have been updated.
- [ ] There are no circuit changes, OR a cryptographer has been assigned for review.
- [ ] I've updated any terraform that needs updating (e.g. environment variables) for deployment.
- [ ] The branch has been rebased against the head of its merge target.
- [ ] I'm happy for the PR to be merged at the reviewer's next convenience.
- [ ] New functions, classes, etc. have been documented according to the doxygen comment format. Classes and structs must have `@brief` describing the intended functionality.
- [ ] If existing code has been modified, such documentation has been added or updated.

> **Note**
> If you are updating the submodule, please make sure you do it in its own _special_ PR and avoid making changes to the submodule as a part of other PRs.
> To update a submodule, you can run the following commands:
> ```console
> $ git submodule update --recursive
> ```
> Alternatively, you can select a particular commit in `barretenberg/aztec3` that you wish to point to:
> ```console
> $ cd barretenberg
> $ git pull origin aztec3        # This will point to the latest commit in `barretenberg/aztec3`
> $ git checkout <commit_hash>    # Use this if you wish to point to a particular commit.
> $ cd ..
> $ git add . && git commit -m <commit_msg>
> $ git push
> ```

### CODE REVIEW

File: cpp/src/aztec3/circuits/rollup/base/native_base_rollup_circuit.cpp [hunk 1]
```diff
@@ -147,8 +147,7 @@ std::array<NT::fr, 3> calculate_new_subtrees(BaseRollupInputs baseRollupInputs,
 
     stdlib::merkle_tree::MemoryTree contracts_tree = stdlib::merkle_tree::MemoryTree(CONTRACTS_SUBTREE_DEPTH);
     stdlib::merkle_tree::MemoryTree commitments_tree = stdlib::merkle_tree::MemoryTree(COMMITMENTS_SUBTREE_DEPTH);
-    // TODO: nullifier tree will be a different tree impl - indexed merkle tree
-    // stdlib::merkle_tree::MemoryTree nullifier_tree = stdlib::merkle_tree::MemoryTree(2);
+    stdlib::merkle_tree::MemoryTree nullifier_tree = stdlib::merkle_tree::MemoryTree(3); // TODO: constant
```
Comment (rahul-kothari) [inline cpp/src/aztec3/circuits/rollup/base/native_base_rollup_circuit.cpp]: nit: can just use `COMMITMENTS_SUBTREE_DEPTH`

File: cpp/src/aztec3/circuits/rollup/base/native_base_rollup_circuit.cpp [hunk 2]
```diff
@@ -275,6 +277,92 @@ void perform_historical_contract_data_tree_membership_checks(BaseRollupInputs ba
...
+
+    // This will increase with every insertion
+    auto new_index = baseRollupInputs.start_nullifier_tree_snapshot.next_available_leaf_index;
+
+    for (size_t i = 0; i < 2; i++) {
+        auto new_nullifiers = baseRollupInputs.kernel_data[i].public_inputs.end.new_nullifiers;
+        for (size_t j = 0; j < 4; j++) {
+
+            // Witness containing index and path
+            auto witness = baseRollupInputs.low_nullifier_membership_witness[i * 4 + j];
```
Comment (rahul-kothari) [inline cpp/src/aztec3/circuits/rollup/base/native_base_rollup_circuit.cpp:310]: nit: probably create a var out of `4i+j`
Also replace 4 with the prev constant?

File: cpp/src/aztec3/circuits/rollup/base/native_base_rollup_circuit.cpp [hunk 3]
```diff
@@ -275,6 +277,92 @@ void perform_historical_contract_data_tree_membership_checks(BaseRollupInputs ba
...
+            // Preimage of the lo-index required for a non-membership proof
+            auto low_nullifier_preimage = baseRollupInputs.low_nullifier_leaf_preimages[i * 4 + j];
+            // Newly created nullifier
+            auto nullifier = new_nullifiers[j];
+
+            // assert that the low_nullifier provided is the correct one by performing two range checks
+            auto is_less_than_nullifier = low_nullifier_preimage.leaf_value < nullifier;
+            auto is_next_greater_than = low_nullifier_preimage.next_value > nullifier;
+            if (!is_less_than_nullifier || !is_next_greater_than) {
```
Comment (rahul-kothari) [inline cpp/src/aztec3/circuits/rollup/base/native_base_rollup_circuit.cpp:319]: I feel like `!(is_less_than_nullifier && is_next_greater_than)` is more readable

File: cpp/src/aztec3/circuits/rollup/base/native_base_rollup_circuit.cpp [hunk 4]
```diff
@@ -196,9 +198,10 @@ std::array<NT::fr, 2> calculate_calldata_hash(BaseRollupInputs baseRollupInputs,
         // Nullifiers
         auto new_nullifiers = baseRollupInputs.kernel_data[i].public_inputs.end.new_nullifiers;
         auto new_commitments = baseRollupInputs.kernel_data[i].public_inputs.end.new_commitments;
-        for (size_t j = 0; j < 4; j++) { // TODO: const
-            calldata_hash_inputs[i * 4 + j] = new_nullifiers[j];
-            calldata_hash_inputs[8 + i * 4 + j] = new_commitments[j];
+        for (size_t j = 0; j < KERNEL_NEW_COMMITMENTS_LENGTH; j++) { // TODO: const
+            calldata_hash_inputs[i * KERNEL_NEW_COMMITMENTS_LENGTH + j] = new_nullifiers[j];
+            calldata_hash_inputs[(KERNEL_NEW_NULLIFIERS_LENGTH * 2) + i * KERNEL_NEW_NULLIFIERS_LENGTH + j] =
```
Comment (rahul-kothari) [inline cpp/src/aztec3/circuits/rollup/base/native_base_rollup_circuit.cpp:203]: can you add a comment to increase readability:
"`KERNEL_NEW_NULLIFIERS_LENGTH * 2` to offset for all nullifiers"
Mayble also add a var for `i * KERNEL_NEW_COMMITMENTS_LENGTH + j` - it is used in other methods too. Can call this nullifier_placeholder_index

File: cpp/src/aztec3/circuits/rollup/base/native_base_rollup_circuit.cpp [hunk 5]
```diff
@@ -275,6 +278,88 @@ void perform_historical_contract_data_tree_membership_checks(BaseRollupInputs ba
         check_membership(leaf, historic_root_witness.leaf_index, historic_root_witness.sibling_path, historic_root);
     }
 }
+
+/**
+ * @brief Check non membership of each of the generated nullifiers in the current tree
+ *
+ * @returns The end nullifier tree root
+ */
+AppendOnlySnapshot check_nullifier_tree_non_membership(BaseRollupInputs baseRollupInputs,
```
Comment (rahul-kothari) [inline cpp/src/aztec3/circuits/rollup/base/native_base_rollup_circuit.cpp:287]: rename this to `check_nullifier_non_membership_and_insert_to_tree()`

File: cpp/src/aztec3/circuits/rollup/base/native_base_rollup_circuit.cpp [hunk 6]
```diff
@@ -140,15 +140,14 @@ AppendOnlySnapshot insert_subtree_to_snapshot_tree(std::array<NT::fr, N> sibling
 std::array<NT::fr, 3> calculate_new_subtrees(BaseRollupInputs baseRollupInputs, std::vector<NT::fr> contract_leaves)
 {
     // Leaves that will be added to the new trees
-    std::array<NT::fr, 8> commitment_leaves; // TODO: use constant
-    std::array<NT::fr, 8> nullifier_leaves;  // TODO: use constant
+    std::array<NT::fr, KERNEL_NEW_COMMITMENTS_LENGTH * 2> commitment_leaves; // TODO: use constant
```
Comment (LHerskind) [inline cpp/src/aztec3/circuits/rollup/base/native_base_rollup_circuit.cpp:143]: remove todos from here and the line below

File: cpp/src/aztec3/constants.hpp [hunk 7]
```diff
@@ -32,6 +32,15 @@ constexpr size_t CONTRACT_TREE_HEIGHT = 4;
 constexpr size_t PRIVATE_DATA_TREE_HEIGHT = 8;
 constexpr size_t NULLIFIER_TREE_HEIGHT = 8;
 
+constexpr size_t CONTRACT_COMMITMENTS_SUBTREE_HEIGHT = 1;
+constexpr size_t CONTRACT_SUBTREE_INCLUSION_CHECK_DEPTH = CONTRACT_TREE_HEIGHT - CONTRACT_COMMITMENTS_SUBTREE_HEIGHT;
```
Comment (rahul-kothari) [inline cpp/src/aztec3/constants.hpp:36]: update in ts

File: cpp/src/aztec3/circuits/rollup/base/native_base_rollup_circuit.cpp [hunk 8]
```diff
@@ -159,12 +158,15 @@ std::array<NT::fr, 3> calculate_new_subtrees(BaseRollupInputs baseRollupInputs,
 
         for (size_t j = 0; j < new_commitments.size(); j++) {
             // todo: batch insert
-            commitments_tree.update_element(i * 4 + j, new_commitments[j]);
+            commitments_tree.update_element(i * KERNEL_NEW_COMMITMENTS_LENGTH + j, new_commitments[j]);
```
Comment (LHerskind) [inline cpp/src/aztec3/circuits/rollup/base/native_base_rollup_circuit.cpp:161]: The subtree to be inserted here need to have the `val`, `next_index`, `next_value`

Comment (cheethas) [reply cpp/src/aztec3/circuits/rollup/base/native_base_rollup_circuit.cpp:161]: will do in another pr

**Pull request merged**